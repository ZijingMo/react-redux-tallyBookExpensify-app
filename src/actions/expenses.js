import uuid from 'uuid'

// These are the 'action' objects down below

// ADD_EXPENSE
export const addExpense = (
  { 
  description = '', 
  note = '', 
  amount = 0, 
  createAt = 0 
  } = {}
) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createAt
  }
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
type: 'REMOVE_EXPENSE',
id
});

// EDIT_EXPENSE
// For the arguments, 'id' is a string number an 'updates' is an object
export const editExpense = (id, updates) => ({
type: 'EDIT_EXPENSE',
id,
updates
});