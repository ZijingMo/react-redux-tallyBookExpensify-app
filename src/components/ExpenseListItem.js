import React from 'react';
import { Link } from 'react-router-dom';

// Destructuring the object to 'dispatch', 'id', 'description', 'amount', 'createAt'
// The component is injected to the 'dispatch' function when we use connect() hoc to wrap it up
const ExpenseListItem = ({ id, description, amount, createAt }) => (
  <div>
    {/* Import 'Link' component to redirect edit page */}
    <Link to={`/edit/${id}`}>
       <h2> { description } </h2> 
    </Link>
    <p> {amount} - {createAt} </p>
  </div>
);

export default ExpenseListItem; 