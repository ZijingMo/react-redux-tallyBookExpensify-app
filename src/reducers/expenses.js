// Expenses Reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id); // If the accessed id match state's id, that id is bye-bye
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }; //Spread syntax updates the state (keep unchanged value and change the fluctuating value)
        } else {
          return expense
        };
      }); 
    default: 
      return state;
  }
};

export default expensesReducer;