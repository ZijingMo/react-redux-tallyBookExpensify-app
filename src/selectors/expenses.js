import moment from 'moment';

// Get visible expenses
// First parameter is only object 'expenses', no destruction
// Second parameter destructs the property from object 'filter'
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const createAtMoment = moment(expense.createAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createAt < b.createAt ? 1 : -1 // It means the most recent created date shows on the top
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1 // It means the most expensive amount shows on the top
    }
  });
}

export default getVisibleExpenses;