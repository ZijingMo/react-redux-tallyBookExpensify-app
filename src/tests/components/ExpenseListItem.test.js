import React from 'react';
//import { Link } from 'react-router-dom';
import { shallow } from 'enzyme';
import ExpenseListItem  from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses'; // This is some dummy data


test('should render ExpenseListItem with expenses', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
  // The meaning of spread '...expenses[0]': 
  // 1. this dummy data has 3 objects in the array, as a result, '{...expenses}' is able to disassemble this dummy data into 3 separately arguments for the component 
  // 2. I chose the first object to test so it looks like '...expenses[0]' 
  expect(wrapper).toMatchSnapshot();
});