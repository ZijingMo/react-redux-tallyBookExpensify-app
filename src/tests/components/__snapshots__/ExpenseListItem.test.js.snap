// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expenses 1`] = `
<div>
  <Link
    to="/edit/axz342"
  >
    <h2>
       
      Candy
       
    </h2>
  </Link>
  <p>
     
    214
     - 
    0
     
  </p>
</div>
`;
